File: src/test/java/li/klass/fhem/service/importexport/ImportExportServiceTest.java
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.service.importexport;

import android.content.Context;

import com.google.common.collect.ImmutableMap;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;

import java.util.Map;

import li.klass.fhem.testutil.MockitoRule;
import li.klass.fhem.util.ApplicationProperties;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.BDDMockito.given;

public class ImportExportServiceTest {
    @Rule
    public MockitoRule mockitoRule = new MockitoRule();

    @Mock
    ApplicationProperties applicationProperties;

    @Mock
    Context applicationContext;

    @InjectMocks
    ImportExportService importExportService;

    @Before
    public void setUp() {
        given(applicationProperties.getApplicationSharedPreferencesName(applicationContext)).willReturn("abc");
    }

    @Test
    @SuppressWarnings("unchecked")
    public void should_match_export_and_import_values() {
        // given
        Map<String, Object> values = ImmutableMap.<String, Object>builder()
                .put("a", "1")
                .put("b", 1)
                .put("c", 1.0)
                .put("f", 1.f)
                .put("d", "1.0")
                .put("e", "anc")
                .put("g", "anc/bas")
                .build();

        // when
        Map<String, Object> converted = (Map<String, Object>) importExportService.toImportValues(importExportService.toExportValues(values));

        // then
        assertThat(converted).isEqualTo(values);
    }

    @Test
    public void should_export_values() {
        // given
        Map<String, Object> values = ImmutableMap.<String, Object>builder()
                .put("a", "1")
                .put("b", 1)
                .put("c", 1.0)
                .put("f", 1.0f)
                .put("d", "1.0")
                .put("e", "anc")
                .put("g", "anc/bas")
                .build();

        // when
        Map<String, String> exported = importExportService.toExportValues(values);

        // then
        assertThat(exported).isEqualTo(ImmutableMap.<String, String>builder()
                .put("a", "1/java.lang.String")
                .put("b", "1/java.lang.Integer")
                .put("c", "1.0/java.lang.Double")
                .put("f", "1.0/java.lang.Float")
                .put("d", "1.0/java.lang.String")
                .put("e", "anc/java.lang.String")
                .put("g", "anc/bas/java.lang.String")
                .build());
    }

    @Test
    @SuppressWarnings("unchecked")
    public void should_import_values() {
        Map<String, String> values = ImmutableMap.<String, String>builder()
                .put("a", "1/java.lang.String")
                .put("b", "1/java.lang.Integer")
                .put("c", "1.0/java.lang.Double")
                .put("f", "1.0/java.lang.Float")
                .put("d", "1.0/java.lang.String")
                .put("e", "anc/java.lang.String")
                .put("g", "anc/bas/java.lang.String")
                .build();

        // when
        Map<String, Object> imported = (Map<String, Object>) importExportService.toImportValues(values);

        // then
        assertThat(imported).isEqualTo(ImmutableMap.<String, Object>builder()
                .put("a", "1")
                .put("b", 1)
                .put("c", 1.0)
                .put("f", 1.0f)
                .put("d", "1.0")
                .put("e", "anc")
                .put("g", "anc/bas")
                .build());
    }
}

File: app/src/test/java/li/klass/fhem/backup/ImportExportServiceTest.kt
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.backup

import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import li.klass.fhem.devices.list.favorites.backend.FavoritesService
import li.klass.fhem.testutil.MockRule
import li.klass.fhem.util.ApplicationProperties
import li.klass.fhem.util.io.FileSystemService
import li.klass.fhem.util.preferences.SharedPreferencesService
import org.assertj.core.api.Assertions.assertThat
import org.junit.Rule
import org.junit.Test

@Suppress("unused")
class ImportExportServiceTest {
    @Rule
    @JvmField
    var mockitoRule = MockRule()

    @MockK
    lateinit var applicationProperties: ApplicationProperties

    @MockK
    lateinit var favoritesService: FavoritesService

    @MockK
    lateinit var sharedPreferencesService: SharedPreferencesService

    @MockK
    lateinit var fileSystemService: FileSystemService

    @InjectMockKs
    lateinit var importExportService: ImportExportService

    @Test
    fun should_match_export_and_import_values() {
        // given
        val values = mapOf<String, Any>(
                "a" to "1",
                "b" to 1,
                "c" to 1.0,
                "f" to 1f,
                "d" to "1.0",
                "e" to "anc",
                "g" to "anc/bas",
                "h" to true
        )

        // when
        val converted = importExportService.toImportValues(importExportService.toExportValues(values)) as Map<*, *>

        // then
        assertThat(converted).isEqualTo(values)
    }

    @Test
    fun should_export_values() {
        // given
        val values = mapOf<String, Any>(
                "a" to "1",
                "b" to 1,
                "c" to 1.0,
                "f" to 1.0f,
                "d" to "1.0",
                "e" to "anc",
                "g" to "anc/bas",
                "h" to true,
                "l" to 1L
        )

        // when
        val exported = importExportService.toExportValues(values)

        // then
        assertThat(exported).isEqualTo(mapOf(
                "a" to "1/java.lang.String",
                "b" to "1/java.lang.Integer",
                "c" to "1.0/java.lang.Double",
                "f" to "1.0/java.lang.Float",
                "d" to "1.0/java.lang.String",
                "e" to "anc/java.lang.String",
                "g" to "anc/bas/java.lang.String",
                "h" to "true/java.lang.Boolean",
                "l" to "1/java.lang.Long"
        ))
    }

    @Test
    fun should_import_values() {
        val values = mapOf(
                "a" to "1/java.lang.String",
                "b" to "1/java.lang.Integer",
                "c" to "1.0/java.lang.Double",
                "f" to "1.0/java.lang.Float",
                "d" to "1.0/java.lang.String",
                "e" to "anc/java.lang.String",
                "g" to "anc/bas/java.lang.String",
                "h" to "true/java.lang.Boolean",
                "l" to "1/java.lang.Long"
        )

        // when
        val imported = importExportService.toImportValues(values) as Map<*, *>

        // then
        assertThat(imported).isEqualTo(mapOf(
                "a" to "1",
                "b" to 1,
                "c" to 1.0,
                "f" to 1.0f,
                "d" to "1.0",
                "e" to "anc",
                "g" to "anc/bas",
                "h" to true,
                "l" to 1L
        ))
    }
}